From 35a2be014507a2fb52d4a4b959942b8849faf4c6 Mon Sep 17 00:00:00 2001
From: Martin Pitt <martin.pitt@ubuntu.com>
Date: Thu, 5 Jan 2012 12:12:33 +0100
Subject: [PATCH] power: Lazily connect to gnome-screensaver

Do not D-BUS activate gnome-screensaver right at startup, this unnecessarily
slows down boot speed. Instead, D-BUS activate it the first time we actually
need it.

Bug: https://bugzilla.gnome.org/show_bug.cgi?id=667345
Bug-Ubuntu: https://launchpad.net/bugs/912186
---
 plugins/power/gsd-power-manager.c |   32 ++++++++++++++++++++------------
 1 files changed, 20 insertions(+), 12 deletions(-)

Index: gnome-settings-daemon-3.3.5/plugins/power/gsd-power-manager.c
===================================================================
--- gnome-settings-daemon-3.3.5.orig/plugins/power/gsd-power-manager.c	2012-02-13 17:31:02.716427557 +0100
+++ gnome-settings-daemon-3.3.5/plugins/power/gsd-power-manager.c	2012-02-13 17:31:02.776427558 +0100
@@ -3358,7 +3358,7 @@
 }
 
 static void
-screensaver_proxy_ready_cb (GObject *source_object,
+sleep_cb_screensaver_proxy_ready_cb (GObject *source_object,
                             GAsyncResult *res,
                             gpointer user_data)
 {
@@ -3370,7 +3370,15 @@
                 g_warning ("Could not connect to gnome-screensaver: %s",
                            error->message);
                 g_error_free (error);
+                return;
         }
+
+        /* Finish the upower_notify_sleep_cb() call by locking the screen */
+        g_debug ("gnome-screensaver activated, doing gnome-screensaver lock");
+        g_dbus_proxy_call (manager->priv->screensaver_proxy,
+                           "Lock",
+                           NULL, G_DBUS_CALL_FLAGS_NONE, -1,
+                           NULL, NULL, NULL);
 }
 
 static void
@@ -3559,6 +3567,17 @@
                                    "Lock",
                                    NULL, G_DBUS_CALL_FLAGS_NONE, -1,
                                    NULL, NULL, NULL);
+        } else {
+                /* connect to the screensaver first */
+                g_dbus_proxy_new_for_bus (G_BUS_TYPE_SESSION,
+                                          G_DBUS_PROXY_FLAGS_DO_NOT_LOAD_PROPERTIES,
+                                          NULL,
+                                          GS_DBUS_NAME,
+                                          GS_DBUS_PATH,
+                                          GS_DBUS_INTERFACE,
+                                          NULL,
+                                          sleep_cb_screensaver_proxy_ready_cb,
+                                          manager);
         }
 }
 
@@ -3727,17 +3746,6 @@
                                   accounts_proxy_ready_cb,
                                   manager);
 
-        /* connect to the screensaver */
-        g_dbus_proxy_new_for_bus (G_BUS_TYPE_SESSION,
-                                  G_DBUS_PROXY_FLAGS_DO_NOT_LOAD_PROPERTIES,
-                                  NULL,
-                                  GS_DBUS_NAME,
-                                  GS_DBUS_PATH,
-                                  GS_DBUS_INTERFACE,
-                                  NULL,
-                                  screensaver_proxy_ready_cb,
-                                  manager);
-
         /* connect to the session */
         g_dbus_proxy_new_for_bus (G_BUS_TYPE_SESSION,
                                   G_DBUS_PROXY_FLAGS_DO_NOT_LOAD_PROPERTIES,
